resolver/resolver-tcp.go

```diff
@@ -386,3 +145,295 @@ func (tr *TCPResolver) Query(ctx context.Context, q *Query) (*RRCache, error) {
+	// refresh dialer to set an authenticated local address
+	// TODO: lock dnsClient (only manager should run at any time, so this should not be an issue)
+	mgr.tr.dnsClient.Dialer = &net.Dialer{
+		LocalAddr: getLocalAddr("tcp"),
+		Timeout:   defaultConnectTimeout,
+		KeepAlive: defaultClientTTL,
+	}
+	// connect
+	c, err := mgr.tr.dnsClient.Dial(mgr.tr.resolver.ServerAddress)
+	if err != nil {
+		log.Debugf("resolver: failed to connect to %s (%s)", mgr.tr.resolver.Name, mgr.tr.resolver.ServerAddress)
+		return false
+	}
+	mgr.conn = c
+	log.Debugf("resolver: connected to %s (%s)", mgr.tr.resolver.Name, mgr.conn.RemoteAddr())
+
+	// reset timer
+	mgr.connTimer.Stop()
+	select {
+	case <-mgr.connTimer.C: // try to empty the timer
+	default:
+	}
+	mgr.connTimer.Reset(mgr.tr.clientTTL)
+
+	// start reader
+	module.StartServiceWorker("dns client reader", 10*time.Millisecond, mgr.msgReader)
+
+	return true
+}
+
+func (mgr *tcpResolverConnMgr) queryHandler() (proceed bool) { //nolint:gocognit
+	var readyToRecycle bool
+
+	for {
+		select {
+		case <-mgr.workerCtx.Done():
+			// module shutdown
+			return false
+
+		case <-mgr.connCtx.Done():
+			// connection error
+			return true
+
+		case <-mgr.connTimer.C:
+			// connection TTL reached, rebuild connection
+			// but handle all in flight queries first
+			readyToRecycle = true
+			// trigger check
+			select {
+			case mgr.responses <- nil:
+			default:
+				// queue is full, check will be triggered anyway
+			}
+
+		case msg := <-mgr.tr.queries:
+			// write query
+			_ = mgr.conn.SetWriteDeadline(time.Now().Add(mgr.tr.dnsClient.WriteTimeout))
+			err := mgr.conn.WriteMsg(msg)
+			if err != nil {
+				if mgr.connClosing.SetToIf(false, true) {
+					mgr.cancelConnCtx()
+					log.Warningf("resolver: write error to %s (%s): %s", mgr.tr.resolver.Name, mgr.conn.RemoteAddr(), err)
+				}
+				return true
+			}
+
+		case msg := <-mgr.responses:
+			if msg != nil { // nil messages only trigger the recycle check
+				// handle query from resolver
+				mgr.tr.Lock()
+				inFlight, ok := mgr.tr.inFlightQueries[msg.Id]
+				if ok {
+					delete(mgr.tr.inFlightQueries, msg.Id)
+				}
+				mgr.tr.Unlock()
+
+				if ok {
+					select {
+					case inFlight.Response <- msg:
+						mgr.failCnt = 0 // reset fail counter
+						// responded!
+					default:
+						// save to cache, if enabled
+						if !inFlight.Query.NoCaching {
+							// persist to database
+							rrCache := inFlight.MakeCacheRecord(msg)
+							rrCache.Clean(600)
+							err := rrCache.Save()
+							if err != nil {
+								log.Warningf(
+									"resolver: failed to cache RR for %s%s: %s",
+									inFlight.Query.FQDN,
+									inFlight.Query.QType.String(),
+									err,
+								)
+							}
+						}
+					}
+				} else {
+					log.Debugf(
+						"resolver: received possibly unsolicited reply from %s (%s): txid=%d q=%+v",
+						mgr.tr.resolver.Name,
+						mgr.conn.RemoteAddr(),
+						msg.Id,
+						msg.Question,
+					)
+				}
```

Comment:
Consider moving that to a dedicated function as it does not depend on loop vars and the possibility of returning allows to outdent multiple blocks:

```golang
func (mgr *tcpResolverConnMgr) handleQueryResponse(msg *dns.Msg) {
	// handle query from resolver
	mgr.tr.Lock()
	inFlight, ok := mgr.tr.inFlightQueries[msg.Id]
	if ok {
		delete(mgr.tr.inFlightQueries, msg.Id)
	}
	mgr.tr.Unlock()

	if !ok {
		log.Debugf(
			"resolver: received possibly unsolicited reply from %s (%s): txid=%d q=%+v",
			mgr.tr.resolver.Name,
			mgr.conn.RemoteAddr(),
			msg.Id,
			msg.Question,
		)
        	return
	}

	select {
	case inFlight.Response <- msg:
		mgr.failCnt = 0 // reset fail counter
		// responded!
	default:	
	}

	// save to cache, if enabled
	if !inFlight.Query.NoCaching {
		// persist to database
		rrCache := inFlight.MakeCacheRecord(msg)
		rrCache.Clean(600)
		err := rrCache.Save()
		if err != nil {
			log.Warningf(
				"resolver: failed to cache RR for %s%s: %s",
				inFlight.Query.FQDN,
				inFlight.Query.QType.String(),
				err,
			)
		}
	}
}
```